runtime/compiler/control/CompilationThread.cpp

```diff
@@ -2521,6 +2527,137 @@ TR::CompilationInfoPerThread* TR::CompilationInfo::getCompInfoForThread(J9VMThre
    return NULL;
    }
 
+int32_t *TR::CompilationInfo::_compThreadActivationThresholds = nullptr;
+int32_t *TR::CompilationInfo::_compThreadSuspensionThresholds = nullptr;
+int32_t *TR::CompilationInfo::_compThreadActivationThresholdsonStarvation = nullptr;
+
+uint32_t TR::CompilationInfo::updateAllowedNumUsableCompThreads(int32_t numCompThreads)
+{
+   if (getPersistentInfo()->getJITaaSMode() == SERVER_MODE)
+      {
+      DEV_DBG("SERVER_MODE numCompThreads %d\n", numCompThreads);
+      _numAllowedUsableCompThreads = ((numCompThreads < 0) ||
+                                      (numCompThreads > MAX_SERVER_USABLE_COMP_THREADS)) ?
+                                      MAX_SERVER_USABLE_COMP_THREADS : numCompThreads;
+      }
+   else
+      {
+      DEV_DBG("CLIENT_MODE numCompThreads %d\n", numCompThreads);
+      _numAllowedUsableCompThreads = ((numCompThreads < 0) ||
+                                      (numCompThreads > MAX_CLIENT_USABLE_COMP_THREADS)) ?
+                                      MAX_CLIENT_USABLE_COMP_THREADS : numCompThreads;
+      }
+
+   _numAllowedTotalCompThreads = _numAllowedUsableCompThreads + _numAllowedDiagnosticCompThreads;
+
+   DEV_DBG("_numAllowedUsableCompThreads %d, _numAllowedDiagnosticCompThreads %d, _numAllowedTotalCompThreads %d\n", _numAllowedUsableCompThreads, _numAllowedDiagnosticCompThreads, _numAllowedTotalCompThreads);
+
+   return _numAllowedUsableCompThreads;
+}
+
+bool
+TR::CompilationInfo::allocateCompilationThreads()
+   {
+   DEV_DBG("_numAllowedUsableCompThreads %d, _numAllowedDiagnosticCompThreads %d, _numAllowedTotalCompThreads %d\n",
+            _numAllowedUsableCompThreads, _numAllowedDiagnosticCompThreads, _numAllowedTotalCompThreads);
+
+   TR::MonitorTable *table = TR::MonitorTable::get();
+   if ((!table) ||
+       (!table->allocInitClassUnloadMonitorHolders(_numAllowedTotalCompThreads)))
+      {
+      return false;
+      }
+
+   if (_compThreadActivationThresholds ||
+       _compThreadSuspensionThresholds ||
+       _compThreadActivationThresholdsonStarvation ||
+       _arrayOfCompilationInfoPerThread)
+      {
+      freeAllCompilationThreads();
```

Comment:
If allocateCompilationThreads can be executed twice, then I would like to prevent that rather than allocating and freeing and then again allocating these arrays

Reply:
Should we assert in debug build?  if (_compThreadActivationThresholds ||
       _compThreadSuspensionThresholds ||
       _compThreadActivationThresholdsonStarvation ||
       _arrayOfCompilationInfoPerThread)
      {
      TR_ASSERT(false, "Compilation threads have been allocated\n");
      return false;
      }

Reply:
recap offline discussion. allocateCompilationThreads() shouldn't be called twice. Add ASSERT_FATAL here.